solutions/0000-0099/0037-sudoku-solver-hard.md

```diff
@@ -0,0 +1,175 @@
+
+
+**Constraints:**
+
+* `board.length == 9`
+* `board[i].length == 9`
+* `board[i][j]` is a digit `1-9` or `'.'`.
+* It is guaranteed that the input board has only one solution.
+
+## Approach 1: Recursion with Backtrack
+
+As we all would have try to solve Sudoku problem in the past, it's `9 x 9` grid with numbers. 
+
+As stated above, a number (in cell) should appear only once in row and column and `3 x 3` block. 
+
+Let's walk through with the given example above, 1st row - `"5","3",".",".","7",".",".",".","."`. The 3rd column (2nd Index) is not filled with any fith any number. 
+
+So what are the possibilities 3rd column can have? Well, It can be from `1-9`, So we use 3 types of checks to place a number on the board. 
+
+1. Check the current row, columns from `1-9`.
+2. Check the current column, rows from `1-9`.
+3. Check the current cell block, which is `3 x 3` grid.
+
+How to define `3 x 3` block start and end range for the current cell we are operating on? Let's say our current cell is $(0, 2)$, we know by seeing the board, it's starts from $(0, 0)$ and end in the range of $(2, 2)$. We can use a simple idea to get the stand and end position, just use `row - row % sqrt(board.length)`, Either we can use $sqrt$ approach or just use 3, since we know here Sudoku is `9 x 9` matrix and any block is size of 3. 
+
+For the above said cell $(0, 2)$, start range of this block is $(0, 0)$, since the start is clear, end is always within `3 x 3` from the start index. 
+
+Now, apply the above said 3 conditions, Try with number $1$, we check from column `1-9`, if the row doesn't have number $1$, then we are clear about row. Next check the column direction (which is row) from `1-9`, we are good again. The Last check to be performed in `3 x 3` block. 
+
+If all $3$ checks is valid, then we can place the number ($1$) in the current cell $(0, 2)$, Now let's move to the next emptied column. In case if we find any invalid cases like, number is already present from in the row or column or `3 x 3`, we try with next number that is $2$. 
+
+So for each cell, we trying with $n$ possibilities, The process continues till for all the rows and columns. If some cases we might end up being not able to place any number and we reached last number that is $9$. **So it means we have placed some wrong order in the past, it requires us to validate the previous placed numbers on the board to solve the current cell problem.** 
+
+This types of situation problems can be solved with backtracking technique. If any of the current operation or action is going to cause future actions then we need to backtrack the previous cases, like placing next number and next number so on to solve the current problem. 
+
+At the end, we would have solved all the cells in the board. 
+
+Time complexity: O(9^n^2) - Exponential time complexity. 
+
+Space complexity: $O(n)$
```

Comment:
Can you double check or explain how to get this.

Reply:
We have total 9 numbers, for any number, n^2 is the possibilities. For every single cell, we are trying out n^2 possibilities, so it's 9^n^2 time complexity. 

Reference - https://www.geeksforgeeks.org/sudoku-backtracking-7/

Reply:
Why SC is O(n)?

Reply:
It's 2D grid, it's a perfect square of 9 x 9, so it's n^2 space complexity. 

Here, m == n, where m - # of rows, n - # of columns.